bokehjs/src/lib/models/ui/pane.ts

```diff
@@ -37,24 +37,44 @@ export class PaneView extends UIElementView {
 
   protected async _update_elements(): Promise<void> {
     const {created} = await this._build_elements()
-    const created_elements = new Set(created)
-
-    // First remove and then either reattach existing elements or render and
-    // attach new elements, so that the order of children is consistent, while
-    // avoiding expensive re-rendering of existing views.
-    for (const element_view of this.element_views) {
-      element_view.el.remove()
-    }
+    const created_views = new Set(created)
+
+    // The newly generated child_views are added to the shadow_el one-by-one
+    // In order to determine the correct ordering we compute the existing
+    // order and then either insert each item before an existing node or append it.
+    // This ensures correct ordering without removing and then re-adding DOM nodes
+    // which can cause issues for certain virtual DOM implementations (e.g. React).
+    const current_views = Array.from(this.shadow_el.children).flatMap(el => {
+      const view = this.element_views.find(view => view.el === el)
+      return view === undefined ? [] : [view]
+    })
 
+    const added = new Set()
     for (const element_view of this.element_views) {
-      const is_new = created_elements.has(element_view)
+      const is_new = created_views.has(element_view)
+      const target = element_view.rendering_target()
 
-      const target = element_view.rendering_target() ?? this.self_target
       if (is_new) {
-        element_view.render_to(target)
+        element_view.render()
+      }
+
+      if (target !== null) {
+        if (!target.contains(element_view.el)) {
+          if (element_view.el.parentNode !== null) {
+            element_view.el.remove()
+          }
```

Comment:
Not needed, if it's already elsewhere in the DOM then it'll be moved on insertion.

```suggestion
```